src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,233 @@
+        @notify_counter += 1
+        spawn watcher(key, files, commands)
+      end
+      @notify_counter_channel.send @notify_counter
+      @notify_counter = @notify_counter_channel.receive
+      sleep
+    end
+
+    private def watcher(key, files, commands)
+      if key != "server"
+        @notify_channel.receive
+      end
+      if files.empty?
+        commands.each do |command|
+          run_command(command, key)
+        end
+      else
+        loop do
+          scan_files(key, files, commands)
+          @watch_running = true if key == "server"
+          sleep 1
+        end
+      end
+    end
+
+    private def scan_files(key, files, commands)
       file_counter = 0
-      Dir.glob(files) do |file|
-        timestamp = get_timestamp(file)
-        if FILE_TIMESTAMPS[file]? != timestamp
-          if @app_running
-            log "File changed: #{file.colorize(:light_gray)}"
-          end
-          FILE_TIMESTAMPS[file] = timestamp
-          file_counter += 1
+      @file_watcher.scan_files(files) do |file|
+        if @watch_running
+          debug "File changed: #{file}"
         end
+        file_counter += 1
       end
       if file_counter > 0
-        log "Watching #{file_counter} files (server reload)..."
-        start_app
+        debug "Watching #{file_counter} #{key} files"
+        kill_processes(key)
+        commands.each do |command|
+          if key == "server" && command == commands.first?
+            run_build_command(command, commands)
+          else
+            run_command(command, key)
+          end
+        end
       end
     end
 
-    private def stop_all_processes
-      log "Terminating app #{project_name}..."
-      @processes.each do |process|
-        process.kill unless process.terminated?
+    private def check_directories
+      Dir.mkdir_p("bin")
+      if !Dir.exists?("lib")
+        error "You need to install dependencies first, execute `shards install`"
+        exit 1
       end
-      processes.clear
     end
 
-    private def create_all_processes
-      process = create_watch_process
-      @processes << process if process.is_a? Process
-      unless @npm_process
-        create_npm_process
-        @npm_process = true
+    private def run_build_command(command, commands)
+      check_directories
+      next_server_commands_range = (1...commands.size)
+      info "Building project #{Helpers.settings.name.colorize(:light_cyan)}"
+      spawn do
+        error_io = IO::Memory.new
+        process = Helpers.run(command, error: error_io)
+        PROCESSES << {process, "server"}
+        loop do
+          if process.terminated?
+            handle_terminaded_process(process, error_io)
+            break
+          end
+          sleep 1
+        end
       end
     end
 
-    private def build_app_process
-      log "Building project #{project_name}..."
-      Amber::CLI::Helpers.run(@build_command)
+    private def handle_terminaded_process(process, error_io)
```

Comment:
Why not a TerminatedProcessHandler classs?

Reply:
I think this useless now, I gonna use parallel processes (like @robacarp  suggested), no channel nor concurrency management would be required :+1: